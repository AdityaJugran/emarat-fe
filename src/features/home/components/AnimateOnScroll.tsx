import styled from 'styled-components';
import { elInViewport } from 'lib/utils';
import { useState, useRef, useEffect } from 'react';

export type PropsType = {
	animationStartClasses?: string;
	children: React.ReactNode;
};

export default function AnimateOnScroll({
	animationStartClasses = 'opacity-0 translate-y-10',
	children,
}: PropsType) {
	const [animationClasses, setAnimationClasses] = useState<string>(
		animationStartClasses
	);

	const rootDivRef = useRef<HTMLDivElement>(null);

	let frameRequestID: number;

	const animate = (currentClasses: string) => {
		let newClasses = currentClasses;

		if (rootDivRef.current) {
			// If element is in veiw then remove the animation classes
			if (elInViewport(rootDivRef.current)) {
				if (currentClasses !== '') {
					newClasses = '';
					setAnimationClasses(newClasses);
				}
			}
			// Otherwise if inital animation classes are not set to intial then do it
			else if (currentClasses !== animationStartClasses) {
				newClasses = animationStartClasses;
				setAnimationClasses(newClasses);
			}
		}
		frameRequestID = requestAnimationFrame(animate.bind(null, newClasses));
	};

	useEffect(() => {
		if (rootDivRef.current) {
			animate(animationStartClasses);
		}

		// Set animation to initial state and cancel animation frame on component unmount
		return () => {
			setAnimationClasses(animationStartClasses);
			if (frameRequestID) cancelAnimationFrame(frameRequestID);
		};
	}, []);

	return (
		<Root ref={rootDivRef} className={`transform ${animationClasses}`}>
			{children}
		</Root>
	);
}

const Root = styled.div`
	transition: transform 2s cubic-bezier(0, 1, 0.3, 1), opacity 0.7s ease-out;
	will-change: transform, opacity;
`;
