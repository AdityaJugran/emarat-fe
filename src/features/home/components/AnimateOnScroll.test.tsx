import AnimateOnScroll from 'features/home/components/AnimateOnScroll';
import renderer from 'react-test-renderer';

it('Animate On Load', () => {
	const tree = renderer
		.create(<AnimateOnScroll>mock child</AnimateOnScroll>)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
