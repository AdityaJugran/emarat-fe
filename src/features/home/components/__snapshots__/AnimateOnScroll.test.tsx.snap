// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animate On Load 1`] = `
<div
  className="sc-bdfBQB bGzMiu transform opacity-0 translate-y-10"
>
  mock child
</div>
`;
